### ISSUE TITLE
Conflict with woocommerce_available_payment_gateways

### ISSUE THREAD

**Issue by BusinessBloomer** [Automattic/woocommerce-payments#4219]
How to reproduce the issue:

1) use woocommerce_available_payment_gateways to disable WooCommerce Payments when a certain shipping method is selected e.g. from https://www.businessbloomer.com/woocommerce-disable-payment-gateway-for-specific-shipping-method/:

```
add_filter( 'woocommerce_available_payment_gateways', 'bbloomer_gateway_disable_for_shipping_rate' );
  
function bbloomer_gateway_disable_for_shipping_rate( $available_gateways ) {
   if ( ! is_admin() ) {
      $chosen_methods = WC()->session->get( 'chosen_shipping_methods' );
      $chosen_shipping = $chosen_methods[0];
      if ( isset( $available_gateways['woocommerce_payments'] ) && 0 === strpos( $chosen_shipping, 'local_pickup' ) ) {
         unset( $available_gateways['woocommerce_payments'] );
      }
   }
   return $available_gateways;
}
```

2) all good, but in a scenario the bug occurs: select the shipping method that hides WooCommerce Payments @ Cart, go to Checkout, so that the shipping method is selected and WooCommerce Payments is hidden (so, woocommerce_available_payment_gateways triggers ok).

Now, switch shipping rate, so that WooCommerce Payments appears again (woocommerce_available_payment_gateways triggers ok again). The Credit Card form does not load: https://www.screencast.com/t/SkvyVs0RXdc6

Basically, it seems that “on checkout load”, if the WooCommerce Gateway is unset, whatever Stripe JS does not run. When you switch shipping and WooCommerce Payments appear back, this JS does not load, and therefore there is no payment card form

Are you able to replicate this?

**Comment by BusinessBloomer** [comment]
p.s. The exact same bug happens with the new WooCommerce PayPal Payments plugin: if on load PayPal is disabled, and I switch shipping rate, PayPal appears but displays no buttons

**Comment by johnaps94** [comment]
Another bug i have found is that when the payment method is getting hidden, the rest of the options chosen result in different option that what is selected.
After testing if for example the woocommerce_payments option is in position 1 and after selecting local_pickup that gets hidden,
if we select for example pay_in_store option (which was in 3rd position before hiding woocommerce_payments option) and is now 2nd position, it results back as if i pressed my 1st position payment option which is credit card...

If i go back then all is normal!
Maybe it has to do something with woocommerce caching

**Comment by MSuar** [comment]
I'm sure it's too late. but the code is limited to an environment outside of the manager! is_admin() and the ajax response is done within the admin. Therefore the answer does not include the code inside if(!is_admin()) statement

**Comment by vbelolapotkov** [comment]
@deepakpathania I see this one is already in your team's maintenance backlog, so let's keep it there.
